engine/runtime-instrument-id-execution/src/main/java/org/enso/interpreter/instrument/IdExecutionInstrument.java

```diff
@@ -90,9 +87,97 @@ public ExecutionEventNode create(EventContext context) {
       return new IdExecutionEventNode(context);
     }
 
-    /**
-     * The execution event node class used by this instrument.
-     */
+    /** Implementation of {@link Info} for the instrumented {@link Node}. */
+    private final class NodeInfo extends Info {
+
+      private final UUID nodeId;
+      private final Object result;
+      private final long elapsedTime;
+      private final MaterializedFrame materializedFrame;
+      private final EnsoRootNode ensoRootNode;
+
+      /**
+       * Create a {@link NodeInfo} for the entered node.
+       *
+       * @param materializedFrame the execution frame
+       * @param node the entered node
+       */
+      public NodeInfo(
+          MaterializedFrame materializedFrame,
+          Node node) {
+        super();
+
+        this.nodeId = getNodeId(node);
+        this.result = null;
+        this.elapsedTime = -1;
+        this.materializedFrame = materializedFrame;
+        this.ensoRootNode = (EnsoRootNode) node.getRootNode();
+      }
+
+      /**
+       * Create a {@link NodeInfo} for the executed node.
+       *
+       * @param nodeId the id of the executed node
+       * @param result the result of the node execution
+       * @param elapsedTime the execution time
+       * @param materializedFrame the execution frame
+       * @param node the executed node
+       */
+      public NodeInfo(
+          UUID nodeId,
+          Object result,
+          long elapsedTime,
+          MaterializedFrame materializedFrame,
+          Node node) {
+        super();
+
+        this.nodeId = nodeId;
+        this.result = result;
+        this.elapsedTime = elapsedTime;
+        this.materializedFrame = materializedFrame;
+        this.ensoRootNode = (EnsoRootNode) node.getRootNode();
+      }
+
+      @Override
+      public UUID getId() {
+        return nodeId;
+      }
+
+      @Override
+      public Object getResult() {
+        return result;
+      }
+
+      @Override
+      public boolean isPanic() {
+        return result instanceof AbstractTruffleException && !(result instanceof DataflowError);
+      }
+
+      @Override
+      public long getElapsedTime() {
+        return elapsedTime;
+      }
+
+      @Override
+      public Object eval(String code) {
+        CallerInfo callerInfo =
+            new CallerInfo(
+                materializedFrame, ensoRootNode.getLocalScope(), ensoRootNode.getModuleScope());
+
+        return evalNode.execute(callerInfo, State.create(EnsoContext.get(null)), Text.create(code));
```

Comment:
Cannot call `evalNode` without entering `CallTarget.call` to achieve reasonable performance. Please try to compare result of some long computation (`fib 50`?) in `main` and in this `eval` mode. I believe you'll see difference in the size of magnitudes.

Reply:
I checked that `fib 50` works fast :+1: